Add compare_set to trampoline class, add typing and formatting

H-Huang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #57203 Update compare_set docs
* **#57191 Add compare_set to trampoline class, add typing and formatting**
* #57175 Update compare_set implementation for FileStore and HashStore

Changed Store::compareSet() to a pure virtual function and added compareSet definition to PythonStore. Rest of changes are from clang-format.

Differential Revision: [D28076557](https://our.internmc.facebook.com/intern/diff/D28076557)

H-Huang committed:
Add compare_set to trampoline class, add typing and formatting

[ghstack-poisoned]

cbalioglu commented on torch/csrc/distributed/c10d/init.cpp:
```diff
@@ -148,6 +162,27 @@ class PythonStore : public ::c10d::Store {
+    // std::vector<uint8_t>. There is no API for directly accessing
+    // the contents of the py::bytes object.
+    std::string str =
+        pybind11::cast<py::bytes>(fn(key, currentValue, newValue));
```
What is the exact return type of `fn`? These two lines seem highly inefficient to me. We copy the return value of `fn` to a temporary `std::string` just for the sake of constructing a `std::vector<>`. If we can have a way to access the data pointer and the size from the return value of `fn`, we can avoid constructing the temporary string object.

H-Huang replied:
For context, PythonStore was originally created to be used as an extension of the store class, so the return value would be whatever the user who is defining the extension implements.

However, it is quite old and I don't believe it is actually being used anymore. Still, I needed to add a compareSet for PythonStore in order to make the Store::compareSet a pure virtual function. I will follow up and find out if it is being used anymore and can potentially remove the entire class in a later PR. I will leave this implementation as is for now.

H-Huang committed:
Update on "Add compare_set to trampoline class, add typing and formatting"


Changed Store::compareSet() to a pure virtual function and added compareSet definition to PythonStore. Rest of changes are from clang-format.

Differential Revision: [D28076557](https://our.internmc.facebook.com/intern/diff/D28076557)

[ghstack-poisoned]

H-Huang committed:
Update on "Add compare_set to trampoline class, add typing and formatting"


Changed Store::compareSet() to a pure virtual function and added compareSet definition to PythonStore. Rest of changes are from clang-format.

Differential Revision: [D28076557](https://our.internmc.facebook.com/intern/diff/D28076557)

[ghstack-poisoned]
